### PR TITLE
Coq 8.15.2 (Coq Platform 2022.04.1)

### PR DESCRIPTION
**Pull request by DonaldKellett**
Related: #1 , codewars/runner#189 

A container image containing the [full package set for Coq Platform 2022.04.0](https://github.com/coq/platform/blob/1af4a9d6fac8f681e0172855fad8805c9877e47a/doc/README~8.15~2022.04.md#coq-platform-2022040-with-coq-8151-full-level). All included libraries are strictly open source.

- Coq 8.15.1 + many mature, curated open-source Coq libraries (incl. Equations and mathcomp)
- coq_codewars 2.0.0
- Ubuntu Jammy base image (slightly larger; [way fewer headaches](http://crunchtools.com/comparison-linux-container-images/))
- Final image size: 3.92 GB - just about 2x the size of the base package set on Jammy (#1 )
- Examples utilize coq_codewars for unit tests
- `bin/run` accepts an environment variable `CONTAINER_ENGINE` (default: `docker`) to support testing the image with Podman

Considering that Coq + all the (50+) curated libraries isn't _that_ large, I say we should include them. A major upside is that the list of packages is standardized so solvers know what to expect and can easily reproduce the exact set of libraries in their local environment, e.g. by running the platform scripts themselves or installing a pre-bundled binary version.

### CODE REVIEW

**Comment by DonaldKellett** [comment]
@kazk I've updated the platform to 2022.04.1 which bumps Coq to 8.15.2.

Feel free to let me know if anything needs to be changed.

File: Dockerfile [hunk 1]
```diff
@@ -1,76 +1,60 @@
...
-RUN apk add --no-cache \
-      build-base \
-      ncurses-dev \
-      ocaml \
-    ;
+RUN useradd -m codewarrior
 COPY --from=builder /opt/coq /opt/coq
-COPY --from=builder /opt/coq_codewars/src /opt/coq_codewars/src
-COPY --from=builder /opt/coq_codewars/theories/Loader.vo /opt/coq_codewars/theories/Loader.vo
+COPY --from=builder /opt/coq_codewars-2.0.0/src /opt/coq_codewars-2.0.0/src
```
Comment (kazk) [inline Dockerfile]: Can you use the same paths without `-2.0.0`?

**Comment by kazk** [comment]
> Final image size: 3.92 GB 

Why is it so huge? Coq 8.12 is 1.11 GB.
Like I wrote before, I'm fine with not using Alpine. But keeping the image size as small as possible is important because it increases the operational cost. It adds up when you have like 60 images to work with.

> A major upside is that the list of packages is standardized so solvers know what to expect and can easily reproduce the exact set of libraries in their local environment

How do you usually manage dependencies in Coq projects?

**Review by kazk — CHANGES_REQUESTED** [review]

**Comment by DonaldKellett** [comment]
> How do you usually manage dependencies in Coq projects?

It used to be that matching versions of libraries had to be manually selected via opam, and it still is for advanced use cases like performing original research with Coq where non-standard (versions of) libraries may be required; however, for standard use cases like education, we now have the option to offload the tedious process of dependency management to the Coq Platform scripts which automate the installation of a standard set of packages.

File: bin/run [hunk 2]
```diff
@@ -1,14 +1,19 @@
...
 
+if [ -z "${CONTAINER_ENGINE:+x}" ]; then
+	CONTAINER_ENGINE=docker
+fi
+
 W=/workspace
+
 # Create container
-C=$(docker container create --rm -w $W ghcr.io/codewars/coq:latest sh -c "coqc Solution.v && coqc -I /opt/coq_codewars/src -Q /opt/coq_codewars/theories CW SolutionTest.v")
+C=$($CONTAINER_ENGINE container create --rm -w $W ghcr.io/codewars/coq:latest sh -c "coqc Solution.v && coqc -I /opt/coq_codewars/src -R /opt/coq_codewars/theories CW SolutionTest.v")
```
Comment (kazk) [inline bin/run]: Why was `-Q` changed to `-R`? `/opt/coq_codewars/theories` doesn't have subdirectories.

File: bin/run [hunk 3]
```diff
@@ -1,14 +1,19 @@
-#!/usr/bin/env bash
+#!/bin/bash
 set -eu
 
+if [ -z "${CONTAINER_ENGINE:+x}" ]; then
+	CONTAINER_ENGINE=docker
```
Comment (kazk) [inline bin/run]:
```suggestion
  CONTAINER_ENGINE=docker
```

File: Dockerfile [hunk 4]
```diff
@@ -1,76 +1,61 @@
...
-    opam install -j 4 coq.8.12.0; \
-    opam pin add coq 8.12.0; \
-    opam install \
-      -j 4 \
-      coq-equations.1.2.3+8.12 \
-      coq-mathcomp-ssreflect.1.11.0 \
-    ;
+    echo "" > stdin; \
+    echo "d" >> stdin; \
+    ./coq_platform_make.sh < stdin;
```
Comment (kazk) [inline Dockerfile]: Let's create a custom `package-pick.sh` based on https://github.com/coq/platform/blob/main/package_picks/package-pick-8.15%7E2022.04.sh by removing packages like `coqide` that's clearly unnecessary for CodeRunner. We can avoid installing many system dependencies that way.
Users can use the original `package-pick` to set up locally.

Also, run `opam clean` to reduce the size of `/opt/coq` after `./coq_platform_make.sh`.

This should reduce the image size significantly.

File: bin/run (hunk 2 above)
Comment (DonaldKellett) [reply bin/run]: I'm not sure :sweat_smile: 

Let me revert the change and see if it runs properly.

File: Dockerfile (hunk 4 above)
Comment (DonaldKellett) [reply Dockerfile]: Thanks! I've added `opam clean` after running the platform scripts and applied a patch file which comments out the CoqIDE part from the package pick. Let's see if this builds.

**Comment by DonaldKellett** [comment]
I've rebased the image to Alpine 3.16 - let's hope this builds and see how much it can further reduce image size. Unfortunately I had to further remove `coq-stdpp`, `coq-iris` and `coq-iris-heap-lang` from the package pick since `coq-stdpp` failed to build on Alpine and the other two are reverse dependencies.

Once this is merged, we should probably mention the packages removed from the package pick (apart from CoqIDE) somewhere in the docs so there are no surprises for solvers and Kata authors.

**Comment by kazk** [comment]
Coq Platform doesn't seem to support installing system dependencies on Alpine. Ubuntu should be fine.

I'll exclude all packages except for `coq-equations` and `coq-mathcomp-ssreflect` for now, and compare with 8.12.

**Comment by kazk** [comment]
Documentation should list the supported packages, and how you can install the same versions easily with Coq Platform. Kind of similar to how Haskell LTS works.

We can add more packages as needed like other languages.

**Comment by DonaldKellett** [comment]
Noted with thanks.

> Coq Platform doesn't seem to support installing system dependencies on Alpine. Ubuntu should be fine.

Might you be encountering ocaml/opam#5186 by chance?

**Comment by kazk** [comment]
> Might you be encountering https://github.com/ocaml/opam/issues/5186 by chance?

It wasn't Coq Platform. I saw `Error loading shared library libgmp.so.10:` and thought Coq Platform's script failed to install `gmp`, but it's actually caused by us using the multi-stage build and the final image missing the dependencies because it's not installed there.

The current image has

https://github.com/codewars/coq/blob/f35ecb5edb6bfc4d5f2981141ab534c156a515f9/Dockerfile#L60-L64

which was missing in your commit.

**Comment by Zimmi48** [comment]
What is the status of this? Any blockers where external help is needed?

**Comment by kazk** [comment]
No, I just forgot. I'll come back during this week.

**Comment by kazk** [comment]
System dependencies are difficult to manage with multi-stage builds, and can be a pain if we want to install more packages in the future. I'll open a new PR for Coq 8.15.2 without multi-stage build that's much smaller.

Also, even though Coq Platform is convenient, I don't think it makes sense to use it to build our image because we're not installing all the packages. I do want to make it easier for the users to get the same package versions, so we'll install the same package versions as the Coq Platform package set. We'll document supported packages, and the package set.

EDIT: #3

**Pull request closed without merging**